README.md
# neuomorphic_container

A package widget that applies the concept of Neuomorphism in software development using Flutter.

[neuomorphic_container Package Url](https://pub.dev/packages/neuomorphic_container)


## Getting Started
Neumorphic containers pretends to extrude from the background. It’s an elevated form created from the same element of the background. 

The background color of your application or widget hosting the Neuomorphic Container has to match the Neuomorphic Container background color to display properly.

The Neuomorphic Container comes with default values for height,width,intensity,blur and offset parameters. 

The blur and intensity parameters are very delicate and the value selected should fit properly. 
You can decide to use the  default value for blur and intensity at first and if that does not fit your purpose you can play around with other values that suits your narrative.

Intensity requires a value within the range of 0.01 to 0.6 else an exception is raised.

The new version comes with four styles for Neuomorphic Container : Flat, Concave, Convex, Pressed. The Pressed Style currently does not work on flutter web. 

## Usage
To use this package, just import 'package:neuomorphic_container/neuomorphic_container.dart';

Explore the example directory for a sample application of this package widget.

Explore the neuomorphism directory for a code generator on the usage of this package. 
The code generator comes with an APK that can be used to test and play around with values to easily decide what fits your design purpose. You can also run the project on a connected iOS device or simulator.


## Sample Application
![ios](example/assets/images/screenshots/ios.png)

![android](example/assets/images/screenshots/android.png)

## Sample Code Generator
![Neuomorphism](neuomorphism/assets/sc/IMG_7174.GIF)


## Inspiration
[Awwwards UI Trends](https://www.awwwards.com/inspiration/neumorphism-in-user-interfaces)

[Neomorphism](https://neumorphism.io/#55b9f3)

## Support
[Buy me a Coffee](https://www.buymeacoffee.com/urNGRFs)

For help getting started with Flutter, check out
[online documentation](https://flutter.dev/docs), which offers tutorials,
samples, guidance on mobile development, and a full API reference.


CHANGELOG.md
## [2.0.3] - 2020-03-06.

* Making minor changes

## [2.0.2] - 2020-03-06.

* Making minor changes

## [2.0.1] - 2020-03-06.

* Extended the styles to four (Convex, Concave, Flat and pressed), It also allows for custom color application and a code generator to test neoumorphic designs and layouts. 

## [1.0.1] - 2020-01-16.

* A flutter widget container that showcases the application of neuomorphism as a concept in UI design

## [0.0.2] - 2020-01-16.

* A flutter widget container that showcases the application of neuomorphism as a concept in UI design

## [0.0.1] - 2020-01-16.

* A flutter widget container that showcases the application of neuomorphism as a concept in UI design


neuomorphism/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:neuomorphism/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


test/neuomorphic_container_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:neuomorphic_container/neuomorphic_container.dart';

void main() {
  test('adds one to input values', () {
 /*   final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
    expect(() => calculator.addOne(null), throwsNoSuchMethodError);*/
  });
}


lib/innerShadow.dart
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'dart:ui' as ui;

class InnerShadow extends SingleChildRenderObjectWidget {
  const InnerShadow({
    Key key,
    this.colorA,
    this.colorB,
    this.blurA,
    this.blurB,
    this.offsetA,
    this.offsetB,
    Widget child,
  }) : super(key: key, child: child);

  final Color colorA;
  final Color colorB;
  final double blurA;
  final double blurB;
  final Offset offsetA;
  final Offset offsetB;

  @override
  RenderInnerShadow createRenderObject(BuildContext context) {
    return RenderInnerShadow()
      ..shadowColorA = colorA
      ..shadowColorB = colorB
      ..blurRadiusA = blurA
      ..blurRadiusB = blurB
      ..offsetGeometryA = offsetA
      ..offsetGeometryB = offsetB;
  }

  @override
  void updateRenderObject(
      BuildContext context, RenderInnerShadow renderObject) {
    renderObject
      ..shadowColorA = colorA
      ..shadowColorB = colorB
      ..blurRadiusA = blurA
      ..blurRadiusB = blurB
      ..offsetGeometryA = offsetA
      ..offsetGeometryB = offsetB;
  }
}

class RenderInnerShadow extends RenderProxyBox {
  RenderInnerShadow({
    RenderBox child,
  }) : super(child);

  @override
  bool get alwaysNeedsCompositing => child != null;

  Color _colorA;
  Color _colorB;
  double _blurA;
  double _blurB;
  Offset _offsetA;
  Offset _offsetB;

  Color get shadowColorA => _colorA;
  set shadowColorA(Color value) {
    if (_colorA == value) return;
    _colorA = value;
    markNeedsPaint();
  }

  Color get shadowColorB => _colorB;
  set shadowColorB(Color value) {
    if (_colorB == value) return;
    _colorB = value;
    markNeedsPaint();
  }

  double get blurRadiusA => _blurA;
  set blurRadiusA(double value) {
    if (_blurA == value) return;
    _blurA = value;
    markNeedsPaint();
  }

  double get blurRadiusB => _blurB;
  set blurRadiusB(double value) {
    if (_blurB == value) return;
    _blurB = value;
    markNeedsPaint();
  }

  Offset get offsetGeometryA => _offsetA;
  set offsetGeometryA(Offset value) {
    if (_offsetA == value) return;
    _offsetA = value;
    markNeedsPaint();
  }

  Offset get offsetGeometryB => _offsetB;
  set offsetGeometryB(Offset value) {
    if (_offsetB == value) return;
    _offsetB = value;
    markNeedsPaint();
  }

  @override
  void paint(PaintingContext context, Offset offset) {
    if (child != null) {
      var layerPaint = Paint()..color = Color.fromRGBO(209, 238, 238, 1.0);

      // Invert the alpha to compute inner part.
      var invertPaint = Paint()
        ..colorFilter = const ui.ColorFilter.matrix([
          1,
          0,
          0,
          0,
          0,
          0,
          1,
          0,
          0,
          0,
          0,
          0,
          1,
          0,
          0,
          0,
          0,
          0,
          -1,
          255,
        ]);

      var canvas = context.canvas;
      canvas.saveLayer(offset & size, layerPaint);
      context.paintChild(child, offset);
      var shadowPaintA = Paint()
        ..blendMode = ui.BlendMode.srcATop
        ..imageFilter =
            ui.ImageFilter.blur(sigmaX: blurRadiusA, sigmaY: blurRadiusA)
        ..colorFilter = ui.ColorFilter.mode(shadowColorA, ui.BlendMode.srcIn);
      var shadowPaintB = Paint()
        ..blendMode = ui.BlendMode.srcATop
        ..imageFilter =
            ui.ImageFilter.blur(sigmaX: blurRadiusB, sigmaY: blurRadiusB)
        ..colorFilter = ui.ColorFilter.mode(shadowColorB, ui.BlendMode.srcIn);
      canvas.saveLayer(offset & size, shadowPaintA);

      canvas.saveLayer(offset & size, invertPaint);
      canvas.translate(_offsetA.dx, _offsetA.dy);
      context.paintChild(child, offset);
      context.canvas.restore();
      context.canvas.restore();
      context.canvas.restore();
      canvas.saveLayer(offset & size, shadowPaintB);
      canvas.saveLayer(offset & size, invertPaint);
      canvas.translate(_offsetB.dx, _offsetB.dy);
      context.paintChild(child, offset);
      context.canvas.restore();
      context.canvas.restore();
      context.canvas.restore();
    }
  }

  @override
  void visitChildrenForSemantics(RenderObjectVisitor visitor) {
    if (child != null) visitor(child);
  }
}


neuomorphism/lib/main.dart
import 'package:flutter/material.dart';
import 'package:neuomorphic_container/neuomorphic_container.dart';
import 'package:neuomorphism/demo.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Neuomorphic Container',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Demo(),debugShowCheckedModeBanner: false,
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
 Color color;
 double distance;
 double blur;
 double intensity;
 double radius;
 double height;
 double width;

 @override
  void initState() {
    color = Color.fromRGBO(209, 238, 238, 1.0);
    intensity =0.2;
    distance=20;
    radius=30.0;
    height=200.0;
    width=200.0;
    blur=60.0;
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
    body: Container(
      padding: EdgeInsets.all(50.0),
      constraints: BoxConstraints.expand(),
      child: Column(
        children:[
          Text("Flutter Neuomorphic Container",style: TextStyle(fontSize: 35.0,fontWeight: FontWeight.w600,color: Colors.blueGrey),),
          SizedBox(height: 10.0),
          Text("Neuomorphic Container Code Generator",style: TextStyle(fontSize: 25.0,fontWeight: FontWeight.w300,color: Colors.blueGrey),),
          SizedBox(height: 10.0),
        SizedBox(child:  NeuomorphicContainer(
            height: height,
            width: width,
            borderRadius: BorderRadius.circular(radius),
            intensity: intensity,
            offset: Offset(distance,distance),
            color: color,
            blur: blur,
            style: NeuomorphicStyle.Pressed,
          ),height: 300,width: 300,
        )
       /*  SizedBox(child: Row(children: <Widget>[
              NeuomorphicContainer(
                height: height,
                width: width,
                borderRadius: BorderRadius.circular(radius),
                intensity: intensity,
                offset: Offset(distance,distance),
                color: color,
                blur: blur,
                style: NeuomorphicStyle.Pressed,
              ),
               SizedBox(width: 100.0),
               NeuomorphicContainer(
                 height: 400,
                 width: 400,
                 borderRadius: BorderRadius.circular(40.0),
                 intensity: intensity,
                 offset: Offset(distance,distance),
                 color: color,
                 blur: blur,
                 style: NeuomorphicStyle.Flat,
               )
              ],mainAxisAlignment: MainAxisAlignment.center,crossAxisAlignment: CrossAxisAlignment.center,mainAxisSize: MainAxisSize.max,

              ),width: 900,


          )*/
        ]
      ),
    ),
      backgroundColor: color,
    );
  }
}


neuomorphism/lib/demo.dart
import 'package:flutter/material.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter_colorpicker/flutter_colorpicker.dart';
import 'package:neuomorphic_container/neuomorphic_container.dart';
import 'package:url_launcher/url_launcher.dart';







class Demo extends StatefulWidget {
  final String title;
  Demo({Key key, this.title}) : super(key: key);

  @override
  State<StatefulWidget> createState() {
   return _Demo();
  }

}

class _Demo extends State<Demo> with SingleTickerProviderStateMixin {

  Color color;
  Color pickedColor;
  double distance;
  double blur;
  double intensity;
  double radius;
  double height;
  double width;
  NeuomorphicStyle style;
  List<Color> navColors;
  TabController _tabController;

  @override
  void initState() {
    color = Color.fromRGBO(209, 238, 238, 1.0);
    intensity = 0.06;
    distance = 5.0;
    radius = 0.0;
    height = 200.0;
    width = 200.0;
    blur = 3.0;
    style = NeuomorphicStyle.Pressed;
    _tabController = TabController(vsync: this, length: 5, initialIndex: 2);
    navColors = [Colors.grey, Colors.grey, color, Colors.grey, Colors.grey];
    super.initState();
  }

  void changeColor(Color selectedColor) {
    setState(() => pickedColor = selectedColor);
  }


  void _showDialog(BuildContext context) {
    // flutter defined function

    showDialog(
        context: context,
        builder: (BuildContext context) {
          return AlertDialog(
            title: const Text('Pick a color!'),
            content: SingleChildScrollView(
              child: ColorPicker(
                pickerColor: color,
                onColorChanged: changeColor,
                showLabel: true,
                pickerAreaHeightPercent: 0.8,
              ),
            ), actions: <Widget>[
            FlatButton(
              child: const Text('Got it'),
              onPressed: () {
                setState(()
                    {color = pickedColor;navColors[1]=pickedColor;});
                Navigator.of(context).pop();
              },
            ),
          ],
          );
        }
    );
  }

  Widget widgetA() {
    return Container(
        constraints: BoxConstraints.expand(),
        color: CupertinoColors.white,
        padding: EdgeInsets.all(20.0),
        child: RichText(text: TextSpan(text: "NeuomorphicContainer(\nheight:",
            style: TextStyle(color: Colors.black, fontSize: 15.0),
            children: [
              TextSpan(text: " ${height.round()}.0, ",
                  style: TextStyle(
                      color: Colors.black, fontWeight: FontWeight.bold)),
              TextSpan(text: "width:", style: TextStyle(color: Colors.black)),
              TextSpan(text: " ${width.round()}.0, ",
                  style: TextStyle(
                      color: Colors.black, fontWeight: FontWeight.bold)),
              TextSpan(text: "borderRadius: BorderRadius.circular(",
                  style: TextStyle(color: Colors.black)),
              TextSpan(text: "${radius.round()}.0",
                  style: TextStyle(
                      color: Colors.black, fontWeight: FontWeight.bold)),
              TextSpan(text: "),\nintensity:",
                  style: TextStyle(color: Colors.black)),
              TextSpan(text: " ${intensity.toStringAsPrecision(2)}, ",
                  style: TextStyle(
                      color: Colors.black, fontWeight: FontWeight.bold)),
              TextSpan(text: "color:", style: TextStyle(color: Colors.black)),
              TextSpan(text: " $color,\n",
                  style: TextStyle(
                      color: Colors.black, fontWeight: FontWeight.bold)),
              TextSpan(text: "offset: Offset(",
                  style: TextStyle(color: Colors.black)),
              TextSpan(text: "${distance.round()}.0,${distance.round()}.0",
                  style: TextStyle(
                      color: Colors.black, fontWeight: FontWeight.bold)),
              TextSpan(text: "), blur:", style: TextStyle(color: Colors.black)),
              TextSpan(text: " ${blur.round()}.0,\n",
                  style: TextStyle(
                      color: Colors.black, fontWeight: FontWeight.bold)),
              TextSpan(text: "style:", style: TextStyle(color: Colors.black)),
              TextSpan(text: " $style\n",
                  style: TextStyle(
                      color: Colors.black, fontWeight: FontWeight.bold)),
              TextSpan(text: ")", style: TextStyle(color: Colors.black)),
            ])
        ));
  }

  Widget widgetD() {
    return Column(
        children: [
          Row
            (children: [
            Text("   Size: ", style: TextStyle(color: Colors.blueGrey,
                fontSize: 15.0,
                fontWeight: FontWeight.w600)),
            Expanded(child: Slider.adaptive(value: height,
                onChanged: (value) =>
                    setState(() {
                      height = value;
                      width = value;
                      distance = height / 10;
                    }),
                label: "$height",
                min: 10.0,
                max: 300.0,
                activeColor: color == CupertinoColors.white
                    ? Colors.red
                    : color,
                inactiveColor: CupertinoColors.inactiveGray))
          ]),
          Row
            (children: [
            Text("   Radius: ", style: TextStyle(color: Colors.blueGrey,
                fontSize: 15.0,
                fontWeight: FontWeight.w600)),
            Expanded(child: Slider.adaptive(value: radius,
                onChanged: (value) =>
                    setState(() {
                      radius = value;
                    }),
                label: "$radius",
                min: 0.0,
                max: 50.0,
                activeColor: color == CupertinoColors.white
                    ? Colors.red
                    : color,
                inactiveColor: CupertinoColors.inactiveGray))
          ]),
          Row
            (children: [
            Text("   Distance: ", style: TextStyle(color: Colors.blueGrey,
                fontSize: 15.0,
                fontWeight: FontWeight.w600)),
            Expanded(child: Slider.adaptive(value: distance,
                onChanged: (value) =>
                    setState(() {
                      distance = value;
                    }),
                label: "$distance",
                min: 5.0,
                max: 50.0,
                activeColor: color == CupertinoColors.white
                    ? Colors.red
                    : color,
                inactiveColor: CupertinoColors.inactiveGray))
          ]),
          Row
            (children: [
            Text("   Blur: ", style: TextStyle(color: Colors.blueGrey,
                fontSize: 15.0,
                fontWeight: FontWeight.w600)),
            Expanded(child: Slider.adaptive(value: blur,
                onChanged: (value) => setState(() => blur = value),
                label: "$blur",
                min: 1.0,
                max: 100.0,
                activeColor: color == CupertinoColors.white
                    ? Colors.red
                    : color,
                inactiveColor: CupertinoColors.inactiveGray))
          ]),


        ]
    );
  }


  Widget widgetB() {
    return Column(children: [Row(
      children: <Widget>[
        Padding(padding: EdgeInsets.only(left: 15.0), child: Column(
          children: [
            Text("Esan Tomisin", style: TextStyle(color: Colors.black,
                fontWeight: FontWeight.bold,
                fontSize: 20.0), textAlign: TextAlign.left),
           IconButton(icon:ImageIcon(AssetImage("assets/twitter.png"),color: color == CupertinoColors.white ? Colors.red : color,size: 50.0,),onPressed: (){
             try{
               _launchURL("https://twitter.com/OluwatomisinEs2");
             }
             catch(e){
               print (e);
             }
           },)
          ]
          , crossAxisAlignment: CrossAxisAlignment.start,)),
        GestureDetector(child:Container(decoration: BoxDecoration(color: Colors.white,
            borderRadius: BorderRadius.circular(20.0),
            border: Border.all(color: Colors.red)),
          child: Padding(padding: EdgeInsets.all(5.0),
              child: Row(children: [
                Image.asset("assets/coffee.png", height: 30.0, width: 30.0),
                Text("  Buy me a coffee  ", style: TextStyle(color: Colors.red))
              ])),
          margin: EdgeInsets.only(right: 10.0),),onTap:(){
          try{
            _launchURL("https://www.buymeacoffee.com/urNGRFs");
          }
          catch(e){
            print (e);
          }
        } )
      ],
      mainAxisAlignment: MainAxisAlignment.spaceBetween,
      crossAxisAlignment: CrossAxisAlignment.start,
    ),
      Row(children: <Widget>[
        Expanded(child: FlatButton.icon(onPressed: () {
          try{
            _launchURL("https://github.com/Tomison-E/neuomorphic_container");
          }
          catch(e){
            print (e);
          }
        },
            icon: Icon(Icons.star, color: Colors.black45,),
            label: Text("Star this repo"))),
        Expanded(child: FlatButton.icon(onPressed: () {
          try{
            _launchURL("https://pub.dev/packages/neuomorphic_container");
          }
          catch(e){
            print (e);
          }
        },
            icon: Icon(Icons.thumb_up, color: Colors.black45,),
            label: Text("Like this package")))
      ],)
    ], mainAxisAlignment: MainAxisAlignment.spaceAround,);
  }

  Widget widgetC() {
    return Column(
      children: [
        Row
          (children: [
          Text("   Intensity: ", style: TextStyle(color: Colors.blueGrey,
              fontSize: 20.0,
              fontWeight: FontWeight.w600)),
          Expanded(child: Slider.adaptive(value: intensity,
              onChanged: (value) => setState(() => intensity = value),
              label: "$intensity",
              min: 0.01,
              max: 0.6,
              activeColor: color == CupertinoColors.white ? Colors.red : color,
              inactiveColor: CupertinoColors.inactiveGray))
        ]),
        FlatButton.icon(onPressed: () => _showDialog(context),
            icon: Icon(Icons.color_lens,
              color: color == CupertinoColors.white ? Colors.red : color,),
            label: Text(
              "select a color", style: TextStyle(color: Colors.blueGrey),)),
      ], mainAxisAlignment: MainAxisAlignment.center,);
  }


  Widget widgetE() {
    return Row
      (children: [
      SizedBox(width: 10.0),
      RaisedButton(color: color == CupertinoColors.white ? Colors.red : color,
        focusColor: color,
        child: Text("Flat"),
        onPressed: () => setState(() => style = NeuomorphicStyle.Flat),),
      RaisedButton(color: color == CupertinoColors.white ? Colors.red : color,
        focusColor: color,
        child: Text("Concave"),
        onPressed: () => setState(() => style = NeuomorphicStyle.Concave),),
      RaisedButton(color: color == CupertinoColors.white ? Colors.red : color,
        focusColor: color,
        child: Text("Pressed"),
        onPressed: ()=> setState(() =>style = NeuomorphicStyle.Pressed)),
      RaisedButton(color: color == CupertinoColors.white ? Colors.red : color,
        focusColor: color,
        child: Text("Convex"),
        onPressed: () => setState(() => style = NeuomorphicStyle.Convex),),
      SizedBox(width: 10.0),
    ], mainAxisAlignment: MainAxisAlignment.spaceEvenly);
  }

  @override
  Widget build(BuildContext context) {
    return  Scaffold(
      appBar: new AppBar(
        backgroundColor: Colors.transparent,
        elevation: 0.0,
        centerTitle: true,
        leading: new IconButton(
          icon: new Icon(Icons.arrow_back_ios),
          onPressed: () {},
        ),
        actions: <Widget>[
          new IconButton(
            icon: new Icon(
              Icons.close,
              size: 26.0,
            ),
            onPressed: () {},
          ),
        ],
      ),
      body:Column(
                children: [
                  Expanded(child: Center(child: NeuomorphicContainer(
                    height: height,
                    width: width,
                    borderRadius: BorderRadius.circular(radius),
                    intensity: intensity,
                    offset: Offset(distance, distance),
                    color: color,
                    blur: blur,
                    style: style,
                  ))),
                Container(
                      decoration: BoxDecoration(
                          borderRadius: BorderRadius.circular(30.0),
                          color: CupertinoColors.white),
                      height: 300.0,
                      child: Column(
                          children: [
                            SizedBox(height: 30.0),
                            SizedBox(child: TabBarView(
                              controller: _tabController,
                              physics: NeverScrollableScrollPhysics(),
                              children: [
                                widgetB(),
                                widgetC(),
                                widgetA(),
                                widgetD(),
                                widgetE(),
                              ],
                            ), height: 200.0),
                            Expanded(child:
                            Padding(padding: EdgeInsets.only(
                                bottom: 30.0, left: 30.0, right: 30.0),
                                child: Row(
                                  children: <Widget>[
                                    IconButton(icon: Icon(
                                        Icons.person, color: navColors[0]),
                                        onPressed: () =>
                                            setState(() {
                                              _tabController.animateTo(0);
                                              navColors = [
                                                color,
                                                Colors.grey,
                                                Colors.grey,
                                                Colors.grey,
                                                Colors.grey
                                              ];
                                            })),
                                    IconButton(icon: Icon(
                                        Icons.format_color_fill,
                                        color: navColors[1]), onPressed: () =>
                                        setState(() {
                                          _tabController.animateTo(1);
                                          navColors = [
                                            Colors.grey,
                                            color,
                                            Colors.grey,
                                            Colors.grey,
                                            Colors.grey
                                          ];
                                        })),
                                    Container(decoration: BoxDecoration(
                                        color: navColors[2],
                                        shape: BoxShape.circle),
                                        child: IconButton(icon: Icon(
                                            Icons.add, color: Colors.white),
                                            onPressed: () =>
                                                setState(() {
                                                  _tabController.animateTo(2);
                                                  navColors = [
                                                    Colors.grey,
                                                    Colors.grey,
                                                    color,
                                                    Colors.grey,
                                                    Colors.grey
                                                  ];
                                                }))),
                                    IconButton(icon: Icon(Icons.aspect_ratio,
                                        color: navColors[3]), onPressed: () =>
                                        setState(() {
                                          _tabController.animateTo(3);
                                          navColors = [
                                            Colors.grey,
                                            Colors.grey,
                                            Colors.grey,
                                            color,
                                            Colors.grey
                                          ];
                                        })),
                                    IconButton(icon: Icon(
                                        Icons.settings, color: navColors[4]),
                                        onPressed: () =>
                                            setState(() {
                                              _tabController.animateTo(4);
                                              navColors = [
                                                Colors.grey,
                                                Colors.grey,
                                                Colors.grey,
                                                Colors.grey,
                                                color
                                              ];
                                            }))
                                  ],
                                  mainAxisAlignment: MainAxisAlignment
                                      .spaceBetween,
                                )))
                          ])
                  )
                ]
            ),backgroundColor: color,

    );
  }

  _launchURL(String url) async {
    if (await canLaunch(url)) {
      await launch(url);
    } else {
      throw 'Could not launch $url';
    }
  }
}


